globus_sdk/config/environments.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import typing
+
+log = logging.getLogger(__name__)
+
+
+class EnvConfig:
+    envname: str
+    domain: str
+    no_dotapi: typing.List[str] = ["auth"]
+
+    # this same dict is inherited (and therefore shared!) by all subclasses
+    _registry: typing.Dict[str, typing.Type["EnvConfig"]] = {}
+
+    # this is an easier hook to use than metaclass definition -- register every subclass
+    # in this dict automatically
+    #
+    # as a result, anyone can define
+    #
+    #       class BetaEnv(EnvConfig):
+    #           domain = "beta.foo.bar.example.com"
+    #           envname = "beta"
+    #
+    # and retrieve it with get_config_by_name("beta")
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._registry[cls.envname] = cls
```

Comment:
`__init_subclass__` is a new hook in 3.6, a lot like metaclass `__new__`, but with a simpler signature and mode of use.

Although the registry is not strictly necessary, since we could define a simple dict mapping `{"production": ProdEnvConfig, ...}`, I think the result is really nice. You can look at the new tests for example usage.